src/lib/stores/wallet/wallet.store.ts

```diff
@@ -0,0 +1,127 @@
+import { ethers } from 'ethers';
+import type { Network } from '@ethersproject/networks';
+import Web3Modal from 'web3modal';
+import { writable } from 'svelte/store';
+import { browser } from '$app/environment';
+
+// https://github.com/vitejs/vite/issues/7257#issuecomment-1079579892
+import WalletConnectProvider from '@walletconnect/web3-provider/dist/umd/index.min.js';
+
+const INFURA_ID = 'aadcb5b20a6e4cc09edfdd664ed6334c';
+
+const WEB_3_MODAL_PROVIDER_OPTIONS = {
+  walletconnect: {
+    package: WalletConnectProvider,
+    options: {
+      infuraId: INFURA_ID,
+    },
+  },
+};
+
+type WalletStoreState = {
+  connected: true;
+  address: string;
+  provider: ethers.providers.Web3Provider;
+  signer: ethers.Signer;
+  network: Network;
+};
+
+export default (() => {
+  const state = writable<WalletStoreState | undefined>(undefined);
+
+  let web3Modal: Web3Modal;
+
+  /**
+   * Initialize the store and restore any previously-connected,
+   * cached provider.
+   */
+  async function initialize(): Promise<void> {
+    if (!browser) return;
+
+    web3Modal = new Web3Modal({
+      providerOptions: WEB_3_MODAL_PROVIDER_OPTIONS,
+      cacheProvider: true,
+    });
+
+    if (web3Modal.cachedProvider) {
+      const instance = await web3Modal.connect();
+      const provider = new ethers.providers.Web3Provider(instance);
+
+      _attachListeners(instance);
+      await _setConnectedState(provider);
+    }
+  }
+
+  /**
+   * Trigger Web3Modal and let the user connect their wallet.
+   */
+  async function connect(): Promise<void> {
+    if (!browser) throw new Error('Can only connect client-side');
+
+    const instance = await web3Modal.connect();
+
+    const provider = new ethers.providers.Web3Provider(instance);
+
+    _attachListeners(instance);
+    await _setConnectedState(provider);
+  }
+
+  /**
+   * Completely clear the store and drop any cached providers from
+   * Web3Modal.
+   */
+  function disconnect(): void {
+    _clear();
+  }
+
+  async function _setConnectedState(provider: ethers.providers.Web3Provider): Promise<void> {
+    const accounts = await provider.listAccounts();
+    const signer = provider.getSigner();
+
+    state.set({
+      connected: true,
+      address: accounts[0],
+      provider,
+      signer,
+      network: await provider.getNetwork(),
+    });
+  }
+
+  function _clear() {
```

Comment:
Only on phone right now so can't check, but isn't that what "clearCachedProvider" does? At least in my testing, after calling that, the connection is not restored after a reload.

Reply:
Just tried with walletconnect on my phone, and it seems to work properly. After clicking disconnect, it doesn't restore on reload. Maybe they patched that.

Reply:
AHA, sorry, I misunderstood.

I think that's actually the intended behavior though. The wallet technically stays connected until you explicitly disconnect it from within the wallet itself (in rainbow by going to the QR scanner > connected apps > choose app > disconnect). It's the same with Metamask — the website itself cannot trigger a real disconnect, so unless you go into MetaMask and explicitly disconnect the site, it will automatically reconnect back to the previously selected account when you choose MetaMask again.

I also think this is weird for sure, and I wish there was some standardized way for a site to explicitly disconnect from a wallet, but as-is, it seems like the expected behavior is the above… wdyt?

Reply:
Aha, now I get it. I misunderstood your point.

I've thought about this with Workstreams before too, and it seems like this is actually intended behavior on WalletConnect's part… unless the user explicitly revokes the connection from within the wallet itself, it's "connected". This whole "disconnect button" thing is really just "simulating" a disconnect, but neither WalletConnect nor MetaMask actually provide a way for the site to explicitly disconnect a wallet.

Here's someone from MetaMask's take on this: https://github.com/MetaMask/metamask-extension/issues/8990#issuecomment-931354298

I think this is mega awkward too… so I'm gonna do what you suggest for WalletConnect, but I don't see a way to do something similar with MetaMask & Web3Modal. Without Web3Modal, the way would be to call "eth_requestAccounts" if the user clicks "connect" while the wallet is already connected, which triggers the account selector and allows easy switching. So I guess we can at least make this behave as expected for WalletConnect, but unless we move away from Web3Modal, I don't think we can do the same for MetaMask :/

Reply:
interesting. i think it's moot for metamask because it gives you a method to control your connected accounts in the extension. wallets via WC often do not so I think it's essential to allow a user to re-connect via WC with a different account (for instance switching from rainbow wallet to argent wallet...)

this disconnect vs logout (delete signature) is interesting as well but i think a bit overly granular for nice UX (something i find endemic to metamask's design in general...). i would support connect as "connect wallet + sign" rather than have "connect" and "login/sign"